File: typescript/tests/hospitality_board.test.ts
// hospitality_board.test.ts
// Verifies the CircleAI.Hospitality port: availability (booked + clean),
// checkout marking a room dirty, front-desk notes newest-first.

import { describe, it } from "node:test";
import assert from "node:assert/strict";
import {
  InMemoryHospitalityBoard,
  HospitalityDomainContext,
  hotelRoom,
  guestReservation,
  frontDeskNote,
} from "../src/hospitality/index";

describe("InMemoryHospitalityBoard", () => {
  it("returns rooms that are clean and not booked over the date", () => {
    const b = new InMemoryHospitalityBoard();
    b.addRoom(hotelRoom("r1", "Std", 1000, "ZAR", true));
    b.addRoom(hotelRoom("r2", "Std", 1000, "ZAR", true));
    b.addRoom(hotelRoom("r3", "Dlx", 2000, "ZAR", false)); // dirty
    // r1 booked across the query date.
    b.reserve(guestReservation("res1", "Ann", "r1", new Date("2026-01-01T00:00:00Z"), new Date("2026-01-05T00:00:00Z")));
    const avail = b.availableOn(new Date("2026-01-03T00:00:00Z")).map((r) => r.roomId);
    assert.deepEqual(avail, ["r2"]);
  });

  it("treats checkout day as available (CheckOut is exclusive)", () => {
    const b = new InMemoryHospitalityBoard();
    b.addRoom(hotelRoom("r1", "Std", 1000, "ZAR", true));
    b.reserve(guestReservation("res1", "Ann", "r1", new Date("2026-01-01T00:00:00Z"), new Date("2026-01-05T00:00:00Z")));
    assert.deepEqual(
      b.availableOn(new Date("2026-01-05T00:00:00Z")).map((r) => r.roomId),
      ["r1"],
    );
  });

  it("checkout marks the room dirty only when cleaning is needed", () => {
    const b = new InMemoryHospitalityBoard();
    b.addRoom(hotelRoom("r1", "Std", 1000, "ZAR", true));
    b.reserve(guestReservation("res1", "Ann", "r1", new Date("2026-01-01T00:00:00Z"), new Date("2026-01-05T00:00:00Z")));
    b.checkOut("res1", true);
    assert.equal(b.getRoom("r1")?.isClean, false);
    assert.equal(b.getReservation("res1")?.guestName, "Ann");
  });

  it("checkout throws on unknown reservation", () => {
    const b = new InMemoryHospitalityBoard();
    assert.throws(() => b.checkOut("ghost", true), /Unknown reservation ghost/);
  });

  it("lists front-desk notes newest-first", () => {
    const b = new InMemoryHospitalityBoard();
    b.addNote(frontDeskNote("n1", "res1", "early", new Date("2026-01-01T08:00:00Z")));
    b.addNote(frontDeskNote("n2", "res1", "late", new Date("2026-01-01T20:00:00Z")));
    b.addNote(frontDeskNote("n3", "res2", "other", new Date("2026-01-01T20:00:00Z")));
    assert.deepEqual(
      b.notesFor("res1").map((n) => n.noteId),
      ["n2", "n1"],
    );
  });

  it("domain context exposes prompt + compliance + tools", () => {
    assert.ok(HospitalityDomainContext.systemPromptSnippet.includes("[DOMAIN: Hospitality]"));
    assert.deepEqual(HospitalityDomainContext.complianceFlags, [
      "Tourism_Act",
      "CATHSSETA",
      "Liquor_Act",
      "Health_Regs",
      "POPIA",
    ]);
    assert.deepEqual(HospitalityDomainContext.suggestedTools, [
      "pms_system",
      "analytics",
      "document_editor",
      "reservation_engine",
    ]);
  });
});

File: go/tests/hospitality_board_test.go
// hospitality_board_test.go
//
// Verifies the CircleAI.Hospitality port (hospitality_board.go): room add/get,
// availability on a date (booked + clean filtering), reserve/check-out with room
// cleaning flag, and notes newest-first.

package circleai_test

import (
	"testing"
	"time"

	circleai "github.com/bhengubv/CircleAI/go"
)

func TestHospitality_Availability(t *testing.T) {
	b := circleai.NewInMemoryHospitalityBoard()
	b.AddRoom(circleai.HotelRoom{RoomId: "101", Type: "Std", NightlyRate: circleai.DecimalFromInt(900), Currency: "ZAR", IsClean: true})
	b.AddRoom(circleai.HotelRoom{RoomId: "102", Type: "Std", NightlyRate: circleai.DecimalFromInt(900), Currency: "ZAR", IsClean: true})
	b.AddRoom(circleai.HotelRoom{RoomId: "103", Type: "Dlx", NightlyRate: circleai.DecimalFromInt(1500), Currency: "ZAR", IsClean: false})
	if got, ok := b.GetRoom("101"); !ok || got.Type != "Std" {
		t.Fatalf("get room = %+v ok=%v", got, ok)
	}

	date := time.Date(2026, 7, 10, 0, 0, 0, 0, time.UTC)
	// 101 booked over the date; 103 not clean; only 102 should be available.
	b.Reserve(circleai.GuestReservation{ReservationId: "r1", GuestName: "Sam", RoomId: "101",
		CheckIn: date.AddDate(0, 0, -1), CheckOut: date.AddDate(0, 0, 2)})
	avail := b.AvailableOn(date)
	if len(avail) != 1 || avail[0].RoomId != "102" {
		t.Fatalf("available-on failed: %+v", avail)
	}
}

func TestHospitality_CheckOutAndNotes(t *testing.T) {
	b := circleai.NewInMemoryHospitalityBoard()
	b.AddRoom(circleai.HotelRoom{RoomId: "201", IsClean: true})
	b.Reserve(circleai.GuestReservation{ReservationId: "r1", GuestName: "Sam", RoomId: "201"})
	if got, ok := b.GetReservation("r1"); !ok || got.GuestName != "Sam" {
		t.Fatalf("get reservation = %+v ok=%v", got, ok)
	}
	if err := b.CheckOut("r1", true); err != nil {
		t.Fatalf("checkout: %v", err)
	}
	if room, _ := b.GetRoom("201"); room.IsClean {
		t.Fatalf("room should be flagged not-clean after checkout")
	}
	if err := b.CheckOut("ghost", false); err == nil {
		t.Fatalf("checkout unknown reservation must error")
	}

	now := time.Now().UTC()
	b.AddNote(circleai.FrontDeskNote{NoteId: "n1", ReservationId: "r1", Body: "early", AtUtc: now.Add(-time.Hour)})
	b.AddNote(circleai.FrontDeskNote{NoteId: "n2", ReservationId: "r1", Body: "late", AtUtc: now})
	notes := b.NotesFor("r1")
	if len(notes) != 2 || notes[0].NoteId != "n2" {
		t.Fatalf("notes newest-first failed: %+v", notes)
	}
}
